test/src/extern/CProducerTest.cpp

```diff
@@ -0,0 +1,229 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "string.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+#include "CMessage.h"
+#include "CProducer.h"
+#include "CSendResult.h"
+
+#include "AsyncCallback.h"
+#include "DefaultMQProducer.h"
+#include "MQMessage.h"
+#include "MQMessageQueue.h"
+#include "MQSelector.h"
+#include "SendResult.h"
+#include "SessionCredentials.h"
+
+using std::string;
+
+using ::testing::_;
+using ::testing::InitGoogleMock;
+using ::testing::InitGoogleTest;
+using testing::Mock;
+using testing::Return;
+
+using rocketmq::DefaultMQProducer;
+using rocketmq::elogLevel;
+using rocketmq::MessageQueueSelector;
+using rocketmq::MQMessage;
+using rocketmq::MQMessageQueue;
+using rocketmq::SendCallback;
+using rocketmq::SendResult;
+using rocketmq::SendStatus;
+using rocketmq::SessionCredentials;
+
+class MockDefaultMQProducer : public DefaultMQProducer {
+   public:
+    MockDefaultMQProducer(const string &groupname) : DefaultMQProducer(groupname) {}
+    MOCK_METHOD0(start, void());
+    MOCK_METHOD0(shutdown, void());
+    MOCK_METHOD2(setLogFileSizeAndNum, void(int, long));
+    MOCK_METHOD1(SetProducerLogLevel, void(elogLevel));
+    MOCK_METHOD2(send, SendResult(MQMessage &, bool));
+    MOCK_METHOD3(send, void(MQMessage &, SendCallback *, bool));
+    MOCK_METHOD2(sendOneway, void(MQMessage &, bool));
+    MOCK_METHOD5(send, SendResult(MQMessage &, MessageQueueSelector *, void *, int, bool));
+};
+
+TEST(cProducer, SendMessageAsync) {
+    MockDefaultMQProducer *mockProducer = new MockDefaultMQProducer("testGroup");
+    CProducer *cProducer = (CProducer *) mockProducer;
+    CMessage *msg = (CMessage *) new MQMessage();
+
+    CSendSuccessCallback cSendSuccessCallback;
+    CSendExceptionCallback cSendExceptionCallback;
+
+    EXPECT_EQ(SendMessageAsync(NULL, NULL, NULL, NULL), NULL_POINTER);
+    EXPECT_EQ(SendMessageAsync(cProducer, NULL, NULL, NULL), NULL_POINTER);
+    EXPECT_EQ(SendMessageAsync(cProducer, msg, cSendSuccessCallback, NULL), NULL_POINTER);
+
+    EXPECT_CALL(*mockProducer, send(_, _)).Times(1);
+    EXPECT_EQ(SendMessageAsync(cProducer, msg, cSendSuccessCallback, cSendExceptionCallback), OK);
+    Mock::AllowLeak(mockProducer);
+}
+
+TEST(cProducer, sendMessageOrderly) {
+    MockDefaultMQProducer *mockProducer = new MockDefaultMQProducer("testGroup");
+    CProducer *cProducer = (CProducer *) mockProducer;
+    CMessage *msg = (CMessage *) new MQMessage();
```

Comment:
Also here,
1. if use `new`, we should free it in the scope.
2. better to use type safe cast.

Reply:
Objects cannot be cast into structures

Reply:
Already processed。delete object